lib/dry/core/memoizable.rb

```diff
@@ -43,47 +43,56 @@ def self.included(klass)
         end
       end
 
-      attr_reader :__memoized__
-
       # @api private
       class Memoizer < Module
-        attr_reader :klass
-        attr_reader :names
-
         # @api private
         def initialize(klass, names)
-          @names = names
-          @klass = klass
-          define_memoizable_names!
+          names.each do |name|
+            define_memoizable(
+              method: klass.instance_method(name)
+            )
+          end
         end
 
         private
 
         # @api private
-        def define_memoizable_names!
-          names.each do |name|
-            meth = klass.instance_method(name)
-
-            if meth.parameters.size > 0
-              define_method(name) do |*args, &block|
-                name_with_args = :"#{name}_#{args.hash}_#{block.hash}"
+        def define_memoizable(method:)
+          module_eval <<~RUBY, __FILE__, __LINE__ + 1
+            def #{method.name}(#{to_declaration(method.parameters)})
+              key = [__method__] + local_variables.map { |var| eval(var.to_s) }
 
-                if __memoized__.key?(name_with_args)
-                  __memoized__[name_with_args]
-                else
-                  __memoized__[name_with_args] = super(*args, &block)
-                end
+              if @__memoized__.key?(key)
+                @__memoized__[key]
+              else
+                @__memoized__[key] = super
               end
+            end
+          RUBY
+        end
+
+        # @api private
+        def to_declaration(params, lookup = params.to_h)
+          params.map do |type, name|
+            case type
+            when :req
+              name
+            when :rest
+              "*#{name}"
+            when :keyreq
+              "#{name}:"
+            when :keyrest
+              "**#{name}"
+            when :block
+              "&#{name}"
+            when :opt
+              lookup.key?(:rest) ? nil : "*args"
+            when :key
+              lookup.key?(:keyrest) ? nil : "**kwargs"
             else
-              define_method(name) do
-                if __memoized__.key?(name)
-                  __memoized__[name]
-                else
-                  __memoized__[name] = super()
-                end
-              end
+              raise NotImplementedError, "type: #{type}, name: #{name}"
```

Comment:
This should never happen, but if it does, this should give us a proper error message.